### DISCUSSION TITLE
Max apps count in config file?

### DISCUSSION THREAD

**Discussion by dw72** [gravitational/teleport#17464] [Q&A]
I try to add about 2000 different apps (Web GUIs for devices) and it seems teleport doesn't read that config. I generate this file from the template, and when I cut it to smaller app count it works ok.
 
 So how many apps can I add to the teleport config?

**Comment by dw72** [comment]
Ok, its looks like the file is read, but teleport does this very very very slow. After the file is read the speed of operations is terrible too :( I suspect that these problems are connected to SQLite on the backend, but is that right? If I change the backend to PostgreSQL it should be faster, but someone can confirm this?

**Reply by webvictim** [reply]
Support for the PostgreSQL backend is actually being removed soon.

What specification is the machine you're running Teleport on?

Could you collect a debug dump? https://goteleport.com/docs/management/admin/troubleshooting/#step-23-generate-a-debug-dump

**Comment by dw72** [comment]
Hi @webvictim,

The machine has 4 vCPU and 8GB of RAM, below is screen from `top` during teleport load of apps (after start but app list is still empty).

![image](https://user-images.githubusercontent.com/1017175/196389389-fc22a500-c0ea-45d5-a6a7-0f7bfd730e19.png)

Here is debug dump: [teleport-dump.txt](https://github.com/gravitational/teleport/files/9808767/teleport-dump.txt)

**Reply by dw72** [reply]
I enabled Write-Ahead Logging on SQLite and now after teleport is started and configuration read it's working ok. But the configuration reading after the start is still very very very slow.

**Reply by webvictim** [reply]
Thanks.

There are two other things you might be able to try:

- running multiple Teleport processes with a smaller set of apps configured on each (and connecting each process back to your main proxy using `auth_servers: ['teleport.example.com:443']`)

- using dynamic app registration (https://goteleport.com/docs/application-access/guides/dynamic-registration/)

**Comment by dw72** [comment]
BTW @webvictim, which backend other than SQLite do you recommend for use on-premise in case of retiring PostgreSQL/CockroachDB?

**Reply by webvictim** [reply]
The only other option is `etcd`

**Reply by dw72** [reply]
But if I use `etcd` where audit events will/should be stored?

The documentation says
> This will allow you to have multiple auth servers in your cluster for an High Availability deployment, but it will not also store Teleport audit events for you in the same way that [DynamoDB](https://goteleport.com/docs/reference/backends/#dynamodb) or [Firestore](https://goteleport.com/docs/reference/backends/#firestore) will.

**Reply by webvictim** [reply]
Correct. In this situation audit events will be stored locally on the disk of each auth server and you would need to forward them to a central SIEM solution for aggregation - perhaps using the [fluentd event forwarder](https://goteleport.com/docs/management/guides/fluentd/).

**Reply by dw72** [reply]
Hi @webvictim, I moved from SQLite to Etcd as a backend, 

```
version: v2
teleport:
  nodename: myhost
  data_dir: /var/lib/teleport
  storage:
    type: etcd
    peers: ["https://10.174.44.52:2379"]
    tls_cert_file: /etc/etcd/certs/teleport.pem
    tls_key_file: /etc/etcd/certs/teleport.key
    tls_ca_file: /etc/etcd/certs/etcd-ca.pem
    prefix: /teleport/
    insecure: false
    etcd_max_client_msg_size_bytes: 15728640
auth_service:
  enabled: "yes"

  ...

```

but in the log still see these lines

```
Nov  8 12:44:16 myhost teleport[30598]: 2022-11-08T12:44:16+01:00 WARN [SQLITE]    "SLOW TRANSACTION: 1.0406859s, goroutine 4082 [running]:\nruntime/debug.Stack()\n\truntime/debug/stack.go:24 +0x65\ngithub.com/gravitational/teleport/lib/backend/lite.(*Backend).inTransaction.func1()\n\tgithub.com/gravitational/teleport/lib/backend/lite/lite.go:938 +0x5b\ngithub.com/gravitational/teleport/lib/backend/lite.(*Backend).inTransaction(0xc000ccc210, {0x7301808, 0xc000078078}, 0xc0043975a8)\n\tgithub.com/gravitational/teleport/lib/backend/lite/lite.go:989 +0x230\ngithub.com/gravitational/teleport/lib/backend/lite.(*Backend).Get(0xc000ccc210, {0x7301808, 0xc000078078}, {0xc000d7fb18, 0x11, 0x18})\n\tgithub.com/gravitational/teleport/lib/backend/lite/lite.go:678 +0x105\ngithub.com/gravitational/teleport/lib/auth.(*ProcessStorage).GetState(0xc001022c20, {0x60e5734, 0x3})\n\tgithub.com/gravitational/teleport/lib/auth/state.go:89 +0x1d3\ngithub.com/gravitational/teleport/lib/service.(*TeleportProcess).getRotation(0x8?, {0x60e5734?, 0x0?})\n\tgithub.com/gravitational/teleport/lib/service/service.go:2084 +0x25\ngithub.com/gravitational/teleport/lib/srv/app.(*Server).getRotationState(_)\n\tgithub.com/gravitational/teleport/lib/srv/app/server.go:442 +0x6e\ngithub.com/gravitational/teleport/lib/srv/app.(*Server).getServerInfo(0xc0030988c0, {0x7340898?, 0xc00388b440?})\n\tgithub.com/gravitational/teleport/lib/srv/app/server.go:429 +0x25c\ngithub.com/gravitational/teleport/lib/srv/app.(*Server).getServerInfoFunc.func1()\n\tgithub.com/gravitational/teleport/lib/srv/app/server.go:402 +0x25\ngithub.com/gravitational/teleport/lib/srv.(*Heartbeat).fetch(0xc0047678c0)\n\tgithub.com/gravitational/teleport/lib/srv/heartbeat.go:333 +0x37\ngithub.com/gravitational/teleport/lib/srv.(*Heartbeat).fetchAndAnnounce(0x0?)\n\tgithub.com/gravitational/teleport/lib/srv/heartbeat.go:605 +0x25\ngithub.com/gravitational/teleport/lib/srv.(*Heartbeat).Run(0xc0047678c0)\n\tgithub.com/gravitational/teleport/lib/srv/heartbeat.go:276 +0x8c\ncreated by github.com/gravitational/teleport/lib/srv/app.(*Server).startHeartbeat\n\tgithub.com/gravitational/teleport/lib/srv/app/server.go:379 +0x21c\n." lite/lite.go:938  
```

I miss something? Why do I still see SLOW TRANSACTION messages from SQLite?

**Reply by webvictim** [reply]
It might be worth clearing out the contents of `/var/lib/teleport/backend` just in case that data is still being used rather than `etcd` (even though it shouldn't be)

There is still a much smaller local sqlite database under `/var/lib/teleport/proc` which is used to store the Teleport-issues host certificate for mutual TLS, but I wouldn't expect that to cause any issues.

**Reply by dw72** [reply]
Probably that little database in _proc_ is causing these warnings, but only when I create/update 2100 dynamic device resources.
